packages/graphile-build-pg/src/QueryBuilder.js

```diff
@@ -868,11 +886,16 @@ order by (row_number() over (partition by 1)) desc`; /* We don't need to factor
       this.locks[type] = isDev ? new Error("Initally locked here").stack : true;
       this.compiledData[type] = data;
     } else if (type === "orderBy") {
-      this.compiledData[type] = this.data[type].map(([a, b, c]) => [
-        callIfNecessary(a, context),
-        b,
-        c,
-      ]);
+      // Add distinct to existing order by to avoid `SELECT DISTINCT ON expressions must match initial ORDER BY expressions`
+      // Convert distinct fields to order by applying ascending and using existing order by for matching fields
+      const distinctAsOrderBy = this.data["distinctOn"].map(
+        d => this.data[type].find(([a]) => isEqual(a, d)) ?? [d, true, null]
```

Comment:
Avoid deep comparison, it has performance overhead. Here's a function you can use instead:

```ts
export function arraysMatch<T>(
  array1: ReadonlyArray<T>,
  array2: ReadonlyArray<T>,
  comparator: (val1: T, val2: T) => boolean = (v1, v2) => v1 === v2,
): boolean {
  const l = array1.length;
  if (l !== array2.length) {
    return false;
  }
  for (let i = 0; i < l; i++) {
    if (!comparator(array1[i], array2[i])) {
      return false;
    }
  }
  return true;
}
```

Use it something like this:

```suggestion
        d => this.data[type].find(([a]) => a === d || (Array.isArray(a) && Array.isArray(d) && arraysMatch(a, d)) ?? [d, true, null]
```